extensions/libraries/joomla_entity/src/Fields/Traits/HasFields.php

```diff
@@ -79,50 +97,92 @@ public function fieldValue($id, $default = null, $raw = false)
 			throw new \InvalidArgumentException($msg);
 		}
 
-		return (null === $values[$id]) ? $default : $values[$id];
+		return (null === $values[$id]) ? $default : $values[$id][$raw ? 'rawvalue' : 'value'];
 	}
 
 	/**
-	 * Get all the field values for this entity.
+	 * Retrieve a field value by a given field name
 	 *
-	 * @param   boolean  $raw  Return raw field values
+	 * @param   integer  $name     Name of the field which value we want to retrieve.
+	 * @param   mixed    $default  Value to use as default if value is null
+	 * @param   boolean  $raw      Return raw field value
+	 *
+	 * @return  mixed
+	 *
+	 * @throws  \InvalidArgumentException  If field value is not set
+	 */
+	public function fieldValueByName($name, $default = null, $raw = false)
+	{
+		$values = $this->fieldValues();
+		$id = array_search($name, array_column($values, 'name'));
+
+		if ($id === false)
+		{
+			return $default;
+		}
+
+		return array_values($values)[$id][$raw ? 'rawvalue' : 'value'];
+	}
+
+	/**
+	 * Get all the field values for this entity.
 	 *
 	 * @return  array
 	 */
-	public function fieldValues($raw = false)
+	public function fieldValues()
 	{
-		$values = array();
-		$fields = $this->fields();
+		// Returns the cached array of values
+		if (null !== $this->fieldValues)
+		{
+			return $this->fieldValues;
+		}
+
+		$fields = $this->fields(false, true);
 
 		if ($fields->isEmpty())
 		{
-			return $values;
+			$this->fieldValues = array();
+
+			return $this->fieldValues;
 		}
 
-		foreach ($fields as $field)
+		$this->fieldValues = array();
+
+		/** @var Field $field */
+		$field = $fields->rewind();
+
+		while ($field)
 		{
-			$property = $raw ? 'rawvalue' : 'value';
-			$values[$field->id()] = $field->get($property);
+			$this->fieldValues[$field->id] = array(
+				'name' => $field->get('name'),
+				'value' => $field->has('value') ? $field->get('value') : '',
+				'rawvalue' => $field->has('rawvalue') ? $field->get('rawvalue') : ''
+			);
+
+			/** @var Field $field */
+			$field = $fields->next();
 		}
 
-		return $values;
+		return $this->fieldValues;
 	}
 
 	/**
 	 * Get associated fields.
 	 *
-	 * @param   boolean  $reload  Force data reloading
+	 * @param   boolean  $reload        Force data reloading
+	 * @param   boolean  $attachValues  Whether to attach values to the fields or not
 	 *
 	 * @return  Collection
 	 */
-	public function fields($reload = false)
+	public function fields($reload = false, $attachValues = false)
 	{
-		if ($reload || null === $this->fields)
+		if ($reload || null === $this->fields || (!$this->withValues && $attachValues))
```

Comment:
When you see these things happening in an if statement you need to start asking yourself what are you doing wrong. Definitely not the way to do it. I understand that you want to pass the attachValues parameter but let's find a better way.

Reply:
The idea behind this change is that, since it's now caching the fields, it needs to force a reload if values were not attached the first time and now they are requested.

However and you're right.  This might go in the same direction of having different functions to do the actual load of fields with or without values, without having to force this parameter.

Let me give it a thought and propose something different.